github/table_github_code_owner.go

```diff
@@ -0,0 +1,166 @@
+		// Top columns
+		{Name: "repository_full_name", Type: proto.ColumnType_STRING, Description: "The full name of the repository, including the owner and repo name."},
+		// Other columns
+		{Name: "line", Type: proto.ColumnType_INT, Description: "The rule's line number in the CODEOWNERS file.", Transform: transform.FromField("LineNumber")},
+		{Name: "pattern", Type: proto.ColumnType_STRING, Description: "The pattern used to identify what code a team, or an individual is responsible for"},
+		{Name: "users", Type: proto.ColumnType_JSON, Description: "Users responsible for code in the repo"},
+		{Name: "teams", Type: proto.ColumnType_JSON, Description: "Teams responsible for code in the repo"},
+		{Name: "pre_comments", Type: proto.ColumnType_JSON, Description: "Specifies the comments added above a key."},
+		{Name: "line_comment", Type: proto.ColumnType_STRING, Description: "Specifies the comment following the node and before empty lines."},
+	}
+}
+
+//// TABLE DEFINITION
+
+func tableGithubCodeOwner() *plugin.Table {
+	return &plugin.Table{
+		Name:        "github_code_owner",
+		Description: "Individuals or teams that are responsible for code in a repository.",
+		List: &plugin.ListConfig{
+			Hydrate:           tableGitHubCodeOwnerList,
+			ShouldIgnoreError: isNotFoundError([]string{"404"}),
+			KeyColumns:        plugin.SingleColumn("repository_full_name"),
+		},
+		Columns: gitHubCodeOwnerColumns(),
+	}
+}
+
+// // LIST FUNCTION
+type CodeOwnerRule struct {
+	LineNumber  int
+	Pattern     string
+	Users       []string
+	Teams       []string
+	PreComments []string
+	LineComment string
+}
+
+func tableGitHubCodeOwnerList(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("tableGitHubCodeOwnerList")
+	repoFullName := d.KeyColumnQuals["repository_full_name"].GetStringValue()
+	owner, repoName := parseRepoFullName(repoFullName)
+
+	type CodeOwnerRuleResponse struct {
+		RepositoryFullName string
+		LineNumber         int
+		Pattern            string
+		Users              []string
+		Teams              []string
+		PreComments        []string
+		LineComment        string
+	}
+	getCodeOwners := func(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+		var fileContent *github.RepositoryContent
+		var err error
+
+		client := connect(ctx, d)
+		opt := &github.RepositoryContentGetOptions{}
+		// stop on the first found CODEOWNERS file.
+		// NOTE : a repository can have multiple CODEOWNERS files, even if it's invalid
+		// In that case, GitHub uses precedence over these files in the following order : .github/CODEOWNERS, CODEOWNERS, docs/CODEOWNERS
+		var paths = []string{".github/CODEOWNERS", "CODEOWNERS", "docs/CODEOWNERS"}
+		for _, path := range paths {
+			fileContent, _, _, err = client.Repositories.GetContents(ctx, owner, repoName, path, opt)
+			if err == nil {
+				break
+			}
+		}
+
+		// no CODEOWNERS file
+		if err != nil {
```

Comment:
Instead of handling all errors as the CODEOWNERS file not being present anywhere, in the `for loop` above, is there a way to instead:
- First check if the file being checked exists
  - If it does exist, try to get its contents
    - If we get an error, return that error immediately
  - If it doesn't exist, move onto the next file
- After the entire `for` loop, if we didn't find any CODEOWNERS files, return `[]CodeOwnerRuleResponse{}, nil`

My thought around this is that if there are other errors that API returns, I believe we should handle/return them so the users knows if something else is wrong, e.g., not enough access.

Reply:
Done @cbruno10 !

Github API doesn't provide a way to test if a file exists.

The only way i found : 
- use `client.Repositories.GetContents` on each path.
- If err is not nil and `err.(*github.ErrorResponse).Response.StatusCode ` is different than 404, return that error immediatly with `err.(*github.ErrorResponse).Response.Status`.
- if `StatusCode` is equal to 404, move onto the next file
- if err is nil, use the file content

I think, the logic is similar.

Reply:
Sounds good to me, thanks!